### PR TITLE
feat: add python 3.12

### LINKED ISSUE #1862: Enhancement: Python 3.12
### Summary

3.12.0b4 is underway and the release date for 3.12 quickly approaches.

Release Schedule: https://peps.python.org/pep-0693/
Available Betas/RCs: https://www.python.org/ftp/python/3.12.0/

Tooling is beginning to support this (https://github.com/astral-sh/ruff/commit/8bc7378002d080094d572e83f3a3a69df1023030), I imagine we will have a waiting game for some things.. but it would be nice to get a head start on this once 2.0 is released.

Support 3.12, adding it to our CI matrix version, testing, docs, etc.

## Pre-3.12 Release Pending Upstream Packages

- [x] `greenlet`
  - [x] https://github.com/python-greenlet/greenlet/pull/327
  - [x] Greenlet 3.0 release
- [x] `grpcio`
	- [x] https://github.com/grpc/grpc/issues/33063
- [x] `orjson >= 3.8.10`
    - [x] https://github.com/ijl/orjson/releases/tag/3.8.10

## Post-3.12 Release Pending Upstream Packages

- [x] https://github.com/MagicStack/uvloop/issues/567

 
--- 

<a href="https://polar.sh/litestar-org/litestar/issues/1862">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/1862/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by JacobCoffee**
### Pull Request Checklist

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

### Description
[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

- Add Python 3.12 as a target - [3.12 Release Schedule](https://peps.python.org/pep-0693/#schedule) is slated to arrive within the next 3 months, with RC 1 coming at the first of next week

### Close Issue(s)
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

- Closes #1862

### CODE REVIEW

**Comment by provinzkraut** [comment]
@JacobCoffee what's the progress here? How many libs are still blocking this?

**Comment by JacobCoffee** [comment]
Still the same - grpc.io and greenlet

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -15,8 +15,8 @@ jobs:
 
       - uses: actions/setup-python@v4
         with:
-          python-version: "3.11"
-
+          python-version: "3.12"
+          allow-prereleases: true
```
Comment (sobolevn) [inline .github/workflows/ci.yaml:19]: You can remove this line now :)
3.12.0 is final!

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -34,14 +34,14 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
         pydantic-version: ["1", "2"]
     uses: ./.github/workflows/test.yaml
     with:
-      coverage: ${{ matrix.python-version == '3.11' && matrix.pydantic-version == '2' }}
+      coverage: ${{ matrix.python-version == '3.12' && matrix.pydantic-version == '2' }}
```
Comment (sobolevn) [inline .github/workflows/ci.yaml:41]: I recommend not to use coverage with 3.12 just yet, there are some possible (and known) bugs, which can result in invalid coverage percentage or line numbers. 3.11 is fine.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -108,8 +108,8 @@ jobs:
       - name: Set up Python
         uses: actions/setup-python@v4
         with:
-          python-version: "3.11"
-
+          python-version: "3.12"
+          allow-prereleases: true
```
Comment (sobolevn) [inline .github/workflows/ci.yaml:112]: ^

File: .github/workflows/docs.yaml [hunk 4]
```diff
@@ -14,7 +14,8 @@ jobs:
       - uses: actions/checkout@v4
       - uses: actions/setup-python@v4
         with:
-          python-version: "3.11"
+          python-version: "3.12"
+          allow-prereleases: true
```
Comment (sobolevn) [inline .github/workflows/docs.yaml:18]: ^

File: .github/workflows/publish.yaml [hunk 5]
```diff
@@ -10,16 +10,16 @@ jobs:
...
       - name: Check out repository
         uses: actions/checkout@v4
 
       - name: Set up Python
         uses: actions/setup-python@v4
         with:
-          python-version: "3.11"
-
+          python-version: "3.12"
+          allow-prereleases: true
```
Comment (sobolevn) [inline .github/workflows/publish.yaml:22]: ^

File: .github/workflows/publish.yaml [hunk 6]
```diff
@@ -47,12 +47,11 @@ jobs:
...
         uses: actions/checkout@v4
-
-      - name: Set up python 3.11
+      - name: Set up Python
         uses: actions/setup-python@v4
         with:
-          python-version: "3.11"
-
+          python-version: "3.12"
+          allow-prereleases: true
```
Comment (sobolevn) [inline .github/workflows/publish.yaml:54]: ^

File: .github/workflows/test.yaml [hunk 7]
```diff
@@ -32,6 +32,7 @@ jobs:
         uses: actions/setup-python@v4
         with:
           python-version: ${{ inputs.python-version }}
+          allow-prereleases: true
```
Comment (sobolevn) [inline .github/workflows/test.yaml:35]: ^

**Review by sobolevn — COMMENTED** [review]
Just a friendly reminder not to use 3.12 syntax anywhere just yet :)
Most tools with just fail to parse it, like mypy does :(

Congrats on 3.12 release 🎉

**Comment by JacobCoffee** [comment]
Thanks @sobolevn! I was trying out the gh cli tool and managed to clone the pr and push it into https://github.com/litestar-org/litestar/pull/2396 - i applied your finding there.

**Pull request closed without merging**